src/VCS/Adapter.php

```diff
@@ -0,0 +1,395 @@
+<?php
+
+namespace Utopia\VCS;
+
+use Exception;
+
+abstract class Adapter
+{
+    public const METHOD_GET = 'GET';
+
+    public const METHOD_POST = 'POST';
+
+    public const METHOD_PUT = 'PUT';
+
+    public const METHOD_PATCH = 'PATCH';
+
+    public const METHOD_DELETE = 'DELETE';
+
+    public const METHOD_HEAD = 'HEAD';
+
+    public const METHOD_OPTIONS = 'OPTIONS';
+
+    public const METHOD_CONNECT = 'CONNECT';
+
+    public const METHOD_TRACE = 'TRACE';
+
+    protected bool $selfSigned = true;
+
+    protected string $endpoint;
+
+    /**
+     * Global Headers
+     *
+     * @var array<string, string>
+     */
+    protected $headers = [];
+
+    /**
+     * Get Adapter Name
+     *
+     * @return string
+     */
+    abstract public function getName(): string;
+
+    /**
+     * Get Adapter Type
+     *
+     * @return string
+     */
+    abstract public function getType(): string;
+
+    /**
+     * Initialize Variables
+     *
+     * @param string $installationId
+     * @param string $privateKey
+     * @param string $githubAppId
+     * @return void
+     */
+    abstract public function initializeVariables(string $installationId, string $privateKey, string $githubAppId): void;
+
+    /**
+     * Generate Access Token
+     *
+     * @param string $privateKey
+     * @param string $githubAppId
+     * @return void
+     */
+    abstract protected function generateAccessToken(string $privateKey, string $githubAppId): void;
+
+    /**
+     * Get user
+     *
+     * @return array<mixed>
+     *
+     */
+    abstract public function getUser(string $username): array;
+
+    /**
+     * Get owner name of the installation
+     *
+     * @return string
+     */
+    abstract public function getOwnerName(string $installationId): string;
+
+    /**
+     * List repositories for Git App
+     * @param int $page page number
+     * @param int $per_page number of results per page
+     * @return array<mixed>
+     */
+    abstract public function listRepositories($page, $per_page): array;
+
+    /**
+     * Get latest opened pull request with specific base branch
+     * @return array<mixed>
+     */
+    abstract public function getBranchPullRequest(string $owner, string $repositoryName, string $branch): array;
```

Comment:
```suggestion
    abstract public function getPullRequestFromBranch(string $owner, string $repositoryName, string $branch): array;
```

we already have a `getPullRequest` method , when creating new methods that are related to it, try to use the same prefix. 

And try to group them together in the file